pkg/component/crddeployer/crd.go

```diff
@@ -0,0 +1,155 @@
+type crdDeployer struct {
+	client            client.Client
+	applier           kubernetes.Applier
+	crdNameToManifest map[string]string
+}
+
+// NewCRDDeployer returns a DeployWaiter that can deploy CRDs and wait for them to be ready.
+func NewCRDDeployer(client client.Client, applier kubernetes.Applier, manifests []string) (component.DeployWaiter, error) {
+	crdNameToManifest, err := MakeCRDNameMap(manifests)
+	if err != nil {
+		return nil, err
+	}
+	return &crdDeployer{
+		client:            client,
+		applier:           applier,
+		crdNameToManifest: crdNameToManifest,
+	}, nil
+}
+
+// Deploy deploys the CRDs.
+func (c *crdDeployer) Deploy(ctx context.Context) error {
+	var fns []flow.TaskFn
+
+	for _, resource := range c.crdNameToManifest {
+		fns = append(fns, func(ctx context.Context) error {
+			return c.applier.ApplyManifest(ctx, kubernetes.NewManifestReader([]byte(resource)), kubernetes.DefaultMergeFuncs)
+		})
+	}
+
+	return flow.Parallel(fns...)(ctx)
+}
+
+// Destroy destroys the CRDs.
+func (c *crdDeployer) Destroy(ctx context.Context) error {
+	var fns []flow.TaskFn
+
+	for resourceName, _ := range c.crdNameToManifest {
+		fns = append(fns, func(ctx context.Context) error {
+			crd := &apiextensionsv1.CustomResourceDefinition{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: resourceName,
+				},
+			}
+			return client.IgnoreNotFound(c.client.Delete(ctx, crd))
+		})
+	}
+
+	return flow.Parallel(fns...)(ctx)
+}
+
+// Wait waits for the CRDs to be deployed.
+func (c *crdDeployer) Wait(ctx context.Context) error {
+	return WaitUntilCRDManifestsReady(ctx, c.client, maps.Keys(c.crdNameToManifest))
+}
+
+// WaitCleanup waits for destruction to finish and CRDs to be fully removed.
+func (c *crdDeployer) WaitCleanup(ctx context.Context) error {
+	return WaitUntilCRDManifestsDestroyed(ctx, c.client, maps.Keys(c.crdNameToManifest))
+}
+
+// MakeCRDNameMap returns a map that has the name of the resource as key, and the corresponding manifest as value.
+func MakeCRDNameMap(manifests []string) (map[string]string, error) {
+	crdNameToManifest := make(map[string]string)
+	for _, manifest := range manifests {
+		name, err := GetObjectNameFromManifest(manifest)
+		if err != nil {
+			return nil, err
+		}
+		crdNameToManifest[name] = manifest
+	}
+	return crdNameToManifest, nil
+}
+
+// WaitUntilCRDManifestsReady takes names of CRDs and waits for them to get ready with a timeout of 15 seconds.
+func WaitUntilCRDManifestsReady(ctx context.Context, c client.Client, crdNames []string) error {
+	var fns []flow.TaskFn
+	for _, crdName := range crdNames {
+		fns = append(fns, func(ctx context.Context) error {
+			timeoutCtx, cancel := context.WithTimeout(ctx, CRDWaitTimeout)
+			defer cancel()
+
+			return retry.Until(timeoutCtx, 1*time.Second, func(ctx context.Context) (done bool, err error) {
+				crd := &apiextensionsv1.CustomResourceDefinition{}
+
+				if err := c.Get(ctx, client.ObjectKey{Name: crdName}, crd); err != nil {
```

Comment:
I assume that at least the first Get call will result in a `NotFoundError` when the cache is not refreshed yet. It may be better to treat those as `retry.MinorError(err)`. Could you double check please?

Reply:
Yup, I will check that, but using `retry.MinorError(err)` really seems to be the way to go here :)